### PR TITLE
IOGX Integration

### PR DESCRIPTION
**Pull request by zeme-wana**
As part of the ongoing effort to build a Developer Platform for IOG, the smart-contracts-dev-empowerment team has releasing [IOGX - a Flake Template for Haskell Projects at IOG](https://github.com/input-output-hk/iogx)

_The vision of IOGX is to provide a JSON-like, declarative interface to Nix, so that developers unfamiliar with the language may independently maintain and add to the Nix sources with minimum effort and maximum pleasure._

The goal is to integrate with all Smart Contracts repositories, and maybe more.

There are currently several other repositories using this: marlowe-cardano, antaeus, quickcheck-dynamic, quickcheck-contractmodel, marconi, stablecoin-plutus

As developers nothing changes for you: you will enter nix develop and you will get into a new shell.

There might be some rough edges but we will be prompt to smooth them.

-----

Since this PR was opened, the interface of IOGX changed:

1. `./nix/iogx-config.nix` renamed to -> `./nix/haskell.nix`
    - `iogx-config.nix:haskellCompilers` -> `./nix/haskell.nix:supportedCompilers`
    - `iogx-config.nix:defaultHaskellCompiler` -> `./nix/haskell.nix:defaultCompiler`
    - `iogx-config.nix:shouldCrossCompile` -> `./nix/haskell.nix:enableCrossCompilation`

2. `./nix/haskell-project.nix` renamed to -> `./nix/cabal-project.nix`

3. `./nix/pre-commit-check.nix` renamed to -> `./nix/formatters.nix`

4. `./nix/hydra-jobs.nix` renamed to -> `./nix/ci.nix`

5. `inputs` and `inputs'` have been swapped.
  `inputs` (no prime `'`) is now the original `inputs`, non-desystemised
  `inputs'` (with prime `'`) is now the de-systemized `inputs`
Reason behind this is that `inputs` is universally understood to be the original, unmodified flake inputs.
While `inputs'` is "non-standard" and thus warrants a different name.
This is the same notation used by [flake.parts](https://flake.parts/module-arguments#inputs)

6. The flake exports aliases for your haskell *executables* only in `outputs.apps`.
   For example, `outputs.apps.marlowe-apps-exe-marlowe-finder-ghc8107` also has an 
   alias `outputs.apps.marlowe-finder` (the cabal package name and the `exe` infix have been stripped)
   These can be access as usual with `inputs'.self.apps.marlowe-finder`

## Release Notes

- **Refactor**: Updated the IOGX template by changing the inputs and outputs of the flake in `flake.nix`.
- **New Feature**: Added a new file `.pre-commit-config.yaml` to improve code quality checks before commits.
- **Refactor**: Modified the `runDockerImage` function and added a new `pushDockerImage` function in `nix/docker-files/docker.nix`.
- **Chore**: Removed large blocks of code related to various packages from multiple `.plan.nix` files, simplifying the project structure.

> 🎉 Here's to the changes that came our way,  
> Refactoring code, making our day.  
> With every line removed, a simpler path unfurled,  
> A cleaner, leaner codebase for the world! 🌍🚀

### CODE REVIEW

File: nix/docker-files/docker.nix [hunk 1]
```diff
@@ -87,72 +86,80 @@
...
-            script="docker push $1/$imageName"
-            echo $script >&2
-            eval "$script"
+      '').outPath;
+    };
 
-        '').outPath;
-      };
+    evaluation-test = pkgs.writeText ''
+      ${runDockerImage.program}
```
[the remark this answers is not in the document]
Comment (zeme-wana) [reply nix/docker-files/docker.nix:162]: The latter

**Comment by RSoulatIOHK** [comment]
@coderabbitai How does it look now?

**Comment by zeme-wana** [comment]
Note that the required statues have changed. From
```
ci/hydra-build:required.x86_64-darwin
ci/hydra-build:required.x86_64-linux
```
To
```
ci/hydra-build:x86_64-linux.required
ci/hydra-build:x86_64-darwin.required
```

Once merged, I'll update the required statuses in the Settings

**Comment by zeme-wana** [comment]
CI is green. I'm going to merge @bogdan-manole @RSoulatIOHK

**Pull request merged**